inventory/inventory.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package inventory
+
+import (
+	"time"
+
+	"github.com/mendersoftware/log"
+	"github.com/mendersoftware/mender/client"
+	"github.com/pkg/errors"
+)
+
+type Attribute struct {
+	Name  string      `json:"name"`
+	Value interface{} `json:"value"`
+}
+
+type Data []Attribute
+
+func (id *Data) replaceAttributes(attr []Attribute) error {
+	iMap := make(map[string]Attribute, len(*id))
+	for _, ia := range *id {
+		iMap[ia.Name] = ia
+	}
+
+	for _, ia := range attr {
+		iMap[ia.Name] = ia
+	}
+
+	cnt := 0
+	for _, v := range iMap {
+		if len(*id) > cnt {
+			(*id)[cnt] = v
+		} else {
+			*id = append(*id, v)
+		}
+		cnt++
+	}
+	return nil
+}
+
+type Inventory struct {
+	sendTicker  *time.Ticker
+	freq        time.Duration
+	scriptsPath string
+	submitter   client.InventorySubmitter
+
+	canSend bool
+}
+
+func New(path string, freq time.Duration, iSubmitter client.InventorySubmitter) *Inventory {
+	ticker := time.NewTicker(freq)
+	tickChan := ticker.C
+
+	i := Inventory{
+		sendTicker:  ticker,
+		freq:        freq,
+		scriptsPath: path,
+		submitter:   iSubmitter,
+	}
+
+	go func() {
+		for {
+			select {
+			case <-tickChan:
+				log.Debug("ticker expired; ready to send inventory data")
+				i.canSend = true
```

Comment:
Thanks for the detailed review. In fact maybe I should have written that in this stage PR was more of a POC to figure out if the design (sending inventory in main thread with bit unspecified schedule) makes sense. Of course all the points you are making here are valid and I didn't mean to merge the code without concurrency support. 
To comment a bit more, I agree that some very aggressive compilers might even remove the whole section as there is no side effect of variable assignment in the given go routine. 
What is more, indeed, I've been thinking of using channels to synchronize. It should be pretty easy in this case and this is nice and idiomatic way of synchronizing things in Go.